### PR TITLE
Integrate MongoDB & Mongoose Library

### PR DESCRIPTION
**Pull request by Maelstromeous**

### CODE REVIEW

File: app/src/services/mongo/MongoDBConnection.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
+import mongoose, {Mongoose} from 'mongoose';
+import ApplicationException from '../../exceptions/ApplicationException';
+import {getLogger} from '../../logger';
+import {injectable} from 'inversify';
+import Database, {DatabaseConfig} from '../../config/database';
+
+@injectable()
+export default class MongoDBConnection {
+    private static readonly logger = getLogger('DatabaseConnection');
+    private static isConnected = false;
```
Comment (microwavekonijn) [inline app/src/services/mongo/MongoDBConnection.ts:10]: Static stuff is going against the principle of IoC.

File: app/src/services/mongo/MongoDBConnection.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    private static async connect(): Promise<void> {
+        MongoDBConnection.logger.info('Starting connection...');
+        const config: DatabaseConfig = MongoDBConnection.dbConfig.config;
+
+        const connStr = `mongodb://${config.user}:${config.pass}@${config.host}:${config.port}?authSource=admin`;
+        MongoDBConnection.logger.debug(connStr);
+
+        mongoose.connection.on('connected', function() {
```
Comment (microwavekonijn) [inline app/src/services/mongo/MongoDBConnection.ts:55]: Don't know if the `connection` property is available before the you call `connect`, but the listener should only be added once so maybe add a connection prepare method that is called on the constructor or on initialization.

Also maybe add some of the other listeners that log at different levels(for debugging mostly).

File: app/src/services/mongo/MongoDatabaseConnectionService.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    // eslint-disable-next-line @typescript-eslint/require-await
+    public async boot(): Promise<void> {
+        MongoDatabaseConnectionService.logger.info('Booting Mongo Database Connection Service...');
+    }
+
+    // eslint-disable-next-line @typescript-eslint/require-await
+    public async start(): Promise<void> {
+        MongoDatabaseConnectionService.logger.info('Starting Mongo Database Connection Service...');
+        this.dbClient.getConnection();
```
Comment (microwavekonijn) [inline app/src/services/mongo/MongoDatabaseConnectionService.ts:24]: I think this needs to be in the boot method as the database should be ready to go once the application starts.

File: app/temp/mongo-example.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+    const kittySchema = new mongoose.Schema({
+        name: String,
+        breed: String
+    })
+
+    kittySchema.methods.speak = function() {
+        const greeting = this.name
+            ? `Meow name is ${this.name}, my breed is ${this.breed}`
+            : "I don't have a name!";
```
Comment (microwavekonijn) [inline app/temp/mongo-example.js:17]: Outrageous! I would throw an error.

File: app/src/services/mongo/MongoDatabaseConnectionService.ts (hunk 3 above)
Comment (marci4) [reply app/src/services/mongo/MongoDatabaseConnectionService.ts:24]: Yeah, I think so too. Or we need an additional function to update different values provided by the database

File: app/temp/mongo-example.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
+const mongoose = require('mongoose')
```
Comment (marci4) [inline app/temp/mongo-example.js:1]: I dont think we need this temp file, right?

File: app/src/services/mongo/MongoDBConnection.ts (hunk 1 above)
Comment (Maelstromeous) [reply app/src/services/mongo/MongoDBConnection.ts:10]: Suggestions for fix? Was really struggling with this one when it came to accessing properties within promises

File: app/src/services/mongo/MongoDBConnection.ts (hunk 2 above)
Comment (Maelstromeous) [reply app/src/services/mongo/MongoDBConnection.ts:63]: I think connection events should always be available, it's an event emitter and it's just merely being registered. Think it's ok here.

File: app/src/services/mongo/MongoDatabaseConnectionService.ts (hunk 3 above)
Comment (Maelstromeous) [reply app/src/services/mongo/MongoDatabaseConnectionService.ts]: Moved to boot method, but it doesn't appear to remain within boot. Suggestions welcome

File: app/src/services/mongo/MongoDBConnection.ts (hunk 2 above)
Comment (microwavekonijn) [reply app/src/services/mongo/MongoDBConnection.ts:63]: As long as they are only added once, should be fine

File: app/src/services/mongo/MongoDBConnection.ts (hunk 1 above)
Comment (microwavekonijn) [reply app/src/services/mongo/MongoDBConnection.ts:10]: Will have a look tomorrow, but it is mostly about removing the static keywords so every new instance can have its own state.

File: app/src/services/mongo/MongoDatabaseConnectionService.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    private readonly dbClient: MongoDBConnection;
+
+    constructor(dbClient: MongoDBConnection) {
+        this.dbClient = dbClient;
+    }
+
+    // eslint-disable-next-line @typescript-eslint/require-await
+    public async boot(): Promise<void> {
+        MongoDatabaseConnectionService.logger.info('Booting Mongo Database Connection Service...');
+        this.dbClient.getConnection();
```
Comment (microwavekonijn) [inline app/src/services/mongo/MongoDatabaseConnectionService.ts:19]: Missing await keyword

File: app/src/services/mongo/MongoDBConnection.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        if (MongoDBConnection.connecting) {
+            MongoDBConnection.logger.info('MongoDBConnection is currently connecting, aborting connection.');
+            return false;
+        }
+
+        if (MongoDBConnection.isConnected) {
+            return MongoDBConnection.db;
+        } else {
+            try {
+                void MongoDBConnection.connect();
```
Comment (microwavekonijn) [inline app/src/services/mongo/MongoDBConnection.ts:29]: Void? I think this should also be await

Comment (Maelstromeous) [reply app/src/services/mongo/MongoDBConnection.ts:29]: Have to convert parent function to async which causes issues, think the method needs rewriting

File: app/src/services/mongo/MongoDatabaseConnectionService.ts (hunk 3 above)
Comment (Maelstromeous) [reply app/src/services/mongo/MongoDatabaseConnectionService.ts]: Figured it out

File: app/src/services/mongo/MongoDBConnection.ts (hunk 1 above)
Comment (Maelstromeous) [reply app/src/services/mongo/MongoDBConnection.ts:10]: Proxying to issue for now, merging code as everything else is resolved

**Pull request merged**